messenger.rst

```diff
@@ -344,59 +285,69 @@ instead of a class name:
 
-        # config/packages/messenger.yaml
-        framework:
-            messenger:
-                routing:
-                    'My\Message\ToBeSentToTwoSenders': [your_transport, audit]
+        public function __construct(int $userId)
+        {
+            $this->userId = $userId;
+        }
 
-    .. code-block:: xml
+        public function getUserId(): int
+        {
+            return $this->userId;
+        }
+    }
 
-        <!-- config/packages/messenger.xml -->
-        <?xml version="1.0" encoding="UTF-8" ?>
-        <container xmlns="http://symfony.com/schema/dic/services"
-            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-            xmlns:framework="http://symfony.com/schema/dic/symfony"
-            xsi:schemaLocation="http://symfony.com/schema/dic/services
-                https://symfony.com/schema/dic/services/services-1.0.xsd
-                http://symfony.com/schema/dic/symfony
-                https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+Then, in your handler, you can query for a fresh object::
 
-            <framework:config>
-                <framework:messenger>
-                    <framework:routing message-class="My\Message\ToBeSentToTwoSenders">
-                        <framework:sender service="your_transport"/>
-                        <framework:sender service="audit"/>
-                    </framework:routing>
-                </framework:messenger>
-            </framework:config>
-        </container>
+    // src/MessageHandler/NewUserWelcomeEmailHandler.php
+    namespace App\MessageHandler;
 
-    .. code-block:: php
+    use App\Message\NewUserWelcomeEmail;
+    use App\Repository\UserRepository;
+    use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
 
-        // config/packages/messenger.php
-        $container->loadFromExtension('framework', [
-            'messenger' => [
-                'routing' => [
-                    'My\Message\ToBeSentToTwoSenders' => ['your_transport', 'audit'],
-                ],
-            ],
-        ]);
+    class NewUserWelcomeEmailHandler implements MessageHandlerInterface
+    {
+        private $userRepository;
+
+        public function __construct(UserRepository $userRepository)
+        {
+            $this->userRepository = $userRepository;
+        }
+
+        public function __invoke(NewUserWelcomeEmail $welcomeEmail)
+        {
+            $user = $this->userRepository->find($welcomeEmail->getUserId());
+
+            // ... send an email!
+        }
+    }
+
+This guarantees the entity contains fresh data.
+
+Handling Messages Synchronously
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-By specifying the ``send_and_handle`` option, you can also route a class of messages to a sender
-while still having them passed to their respective handler:
+If a message doesn't :ref:`match any routing rules <messenger-routing>`, it won't
+be sent to any transport and will be handled immediately. In some cases, it's easier
```

Comment:
I think we should give at least one example of "some cases". At the moment, I'm unsure when it's more usefull to define it explicitly.